controllers/machinedeletionremediation_controller.go

```diff
@@ -282,37 +281,43 @@ func (r *MachineDeletionRemediationReconciler) getNodeFromMdr(ctx context.Contex
 	return node, nil
 }
 
-// getMachine retrieves a machine from the cluster based on the remediation.
-// It returns the machine, a boolean indicating whether the machine was expected to be found,
-// and an error if any occurred during the retrieval process.
+// getMachine retrieves a Machine from the cluster based on the remediation.
+// It returns the machine and an error if any occurred during the retrieval process.
 func (r *MachineDeletionRemediationReconciler) getMachine(ctx context.Context, remediation *v1alpha1.MachineDeletionRemediation) (*machinev1beta1.Machine, error) {
-	// If the Machine's Name and Ns come from the related Node, it is expected
-	// to find the Machine in the cluster, while if its Name and Ns come from
-	// CR's annotation, the Machine might have been deleted upon our request.
+	// The Name and Namespace to retrieve the target Machine can come from the following sources:
+	// - the remediation's ownerReference: if the remediation was created by MachineHealthcheck
+	// - the remediation's Node: if the remediation was created by NodeHealthcheck or manually
+	// - the remediation's MachineNameNsAnnotation annotation: once the Machine is found and its Name and Namespace are saved in
+
+	// Try to get first the Machine's data from the remediation's annotation, if any. It means that the Machine was already
+	// been found in a previous cycle and we can use the data to verify if the Machine was deleted upon our request or not
 	machineName, machineNs, err := getRemediationDataFromAnnotation(remediation, MachineNameNsAnnotation)
 	if err != nil {
 		r.Log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", MachineNameNsAnnotation)
 		return nil, unrecoverableError
 	}
 
-	var mustExist bool
-	if machineName == "" {
-		// Remediation does not have the MachineDataAnnotation yet. We will try to get the Machine's data from its Node,
-		// and we expect to find the Node and the Machine in the cluster.
-		mustExist = true
+	// If the Machine's Name is not in the annotation, it means that it must come from the other two sources and in
+	// turns it means that the Machine must exist in the cluster, otherwise an error is returned.
+	mustExist := machineName == ""
+
+	if machineName == "" && len(remediation.GetOwnerReferences()) > 0 {
```

Comment:
`machineName == ""` can be replaced with `mustExist`

Reply:
technically yes, semantically not so much IMHO...

Reply:
I think this pattern is confusing.
```
someBool :=  someEvaluation
if someEvaluation {
// Do something
}
```

When I read this sort of code I immediately wonder why wasn't boolean used in the `if` clause since the exact same evaluation was done one line before.
I suggest either add a comment on `mustExist` that explains that, or use a different name than `mustExist` that will explain it.

Reply:
> or use a different name

+1
maybe something like `isUnhandledMachine`, which explains the situation, and implies indirectly that the machine has to exist

Reply:
I think it's better with the comment rather than without, but it still doesn't explain why we are not using the boolean in the `if` clause.
Let me suggest the following refactoring:

```
//this method will contain all the logic of getting the machineName (annotation/ownerRef or remediation)
machineName, machineNs, err := getRemediationData(remediation, MachineNameNsAnnotation)
...
if err := r.Get(ctx, key, machine); err != nil {
..
  if  isMachineNameExistInAnnotation(remediation, MachineNameNsAnnotation){
  ...
 ```
 
 I think it'll make the code much clearer, one downside is that we might need to get the data from the annotation twice- but IMO it is worth the trade-off.
 
 In any case feel free to use different method name, or disregard completely - I don't think this is major enough to block on.

Reply:
I understand your point. Maybe we can get the same result renaming `mustExist` as you suggested in `isUnhandledMachine` (as suggested by Marc)?

I propose this because I think that `isMachineNameExistInAnnotation` would still need an explanation as well as `mustExist` does. Changing the name I could write something like (kind of pseudo code)

```
name, ns := getDataFromAnnotation()

isUnhandledMachine := name == ""

if isUnhandledMachine {
  // get the Machine from the cluster
}

```

wdyt?

Reply:
In general I think it's a good idea to use `isUnhandledMachine` in the if clause instead of evaluating `name == ""` again,  which I find confusing.
I also think that `isUnhandledMachine` is a better name.

I do think this is still a bit confusing because `isUnhandledMachine` value may not reflect the actual logic.
IIUC We only decide that the machine is unhandled after we also check ownerRef & remediation.
So to put it in other words `isUnhandledMachine` value may equal `true` when logically it is actually false.

Another option is just to rename `isMachineNameExistInAnnotation` method to `isUnhandledMachine`in my original suggestion.

In any case I do think your suggestion is better than the current status and in any case I consider this a NIT so do feel free to implement it as you decide best :smile:

Reply:
Let's merge this now then, I'll follow up with a specific PR :)